windows/.bazelrc

```diff
@@ -3,7 +3,8 @@
 build --define signal_trace=disabled
 build --define hot_restart=disabled
 build --define tcmalloc=disabled
-# Build GRPC without c-ares support -- otherwise it will not link
-build --define grpc_no_ares=true
-build --features=compiler_param_file
 build --workspace_status_command="bash bazel/get_workspace_status"
+
+build --experimental_enable_runfiles
+build --features=compiler_param_file
+build --features=static_link_msvcrt
```

Comment:
Oh I didn't notice this is an official Bazel feature, if that is the case, shouldn't external foreign_cc propagate this correctly without patching CMakeLists files? @irengrig Any thoughts?

Reply:
@lizan you give too much credit to cmake even, which doesn't support it well. (Thus the specific patches for specific dependent projects.)

A truly clever fix would add core support at the cmake project (while fixing .pdb file handling, even), combined with injecting carrying this flag from bazel to the cmake toggle. But that's beyond the remit of this proposed patch.

Reply:
> A truly clever fix would add core support at the cmake project (while fixing .pdb file handling, even), combined with injecting carrying this flag from bazel to the cmake toggle. But that's beyond the remit of this proposed patch.

This is what I'm asking @irengrig, I believe CMake toolchain can handle them, though rules_foreign_cc doesn't do it yet.

Reply:
>shouldn't external foreign_cc propagate this correctly without patching CMakeLists files? @irengrig Any thoughts?

Yep, if it is a good matching, we totally should do it.

Could you elaborate more about the CMake cache values that should be set for what Bazel features? (static_link_msvcrt -> ??? CMAKE_CXX_FLAGS = ...???)

Reply:
Your understanding is correct @lizan. It isn't additive, but we need to drop the /MD[d], swapping it with /MT[d], only in the case of bazel build --features=static_link_msvcrt (in fact it could be helpful to swap /MT[d] to /MD[d] when bazel is not built this way.)

P.s. the /MDd /MTd additional [d] bit simply refers to the debug CRT.

Reply:
FWIW this doesn't block this patch submission... we can certainly expect more from the cmake and bazel projects in the future, but let's please proceed today to work around this shortcoming. The transliteration of /MD->/MT will be a noop when the feature is rolled out, and eventually would obsolete this patch.

Reply:
Where is /MD added? by CMake itself? Seems CMAKE_{C,CXX}_FLAGS can control them, no? Agree the foreign_cc improvement doesn't block the submission, but I'd like to minimize upstream patches we maintain, at least having a path we can get rid of them.

Reply:
Yes, CMake does this, but smashing all the rest of their copts isn't a great approach. I'm happy to champion a consistent fix upstream, but we realistically will have to wait two years for that benefit.

Reply:
rules_foreign_cc already does smashing all the rest of copts and use what is provided to bazel, so it is in the right position to do it.

Reply:
I checked it more carefully now, and from what I see, Bazel takes the information from --features=static_link_msvcrt and transforms it into passing either /MD or /MT flags in the Bazel's supplied C++ flags. For cmake_external, rules_foreign_cc are converting it into the value of CMAKE_CXX_FLAGS_INIT variable - which is the initializer for CMAKE_CXX_FLAGS. As far as I know, this value is combined with build-type value like CMAKE_CXX_FLAGS_RELEASE, 
but nevertheless, CMAKE_CXX_FLAGS value will be used.

Also, if I additionally specify --compilation_mode dbg, I am getting the flag variant /MTd (/MDd) from Bazel.
The question is, do we need to convert users' flags (/MD[d] <=> /MT[d])?

Is that enough what Bazel is already doing?
/cc @hlopko

Reply:
@irengrig the confusing part is that the FLAGS_{BUILDTYPE} are appended to the FLAGS already managed by rules_foreign_cc build_cmake. So Bazel's preferences are crushed by overrides based on CMAKE_BUILD_TYPE.

We believe we have the optimal solution, always set CMAKE_BUILD_TYPE to "Bazel". This means that FLAGS_BAZEL values are appended, but these are empty, so that's a no-op, and the various compile/link settings arriving from rules_foreign_cc are preserved.

Tested on the opt build, doing dbg now, will follow on with fastbuild and propose an additional commit which takes out all of the static build logic patches, but adds some for a serious rules_foreign_cc build quirk. Although I believed defines=[] was propagating, and it does to native cc rules, the build_cmake logic in rules_foreign_cc fails entirely to promote these from the component's dependencies into its CFLAGS. Very ugly to hack around this for the time being, but those hacks are actually useful to the traditional cmake users anyways, so the fixes will be floated upstream. Will demonstrate this tomorrow.

Reply:
After reviewing opt, dbg and fastbuild results, using a custom "Bazel" CMAKE_ I have two observations...

It's wonderful. Bazel fastbuild already doesn't map to a CMAKE target. On windows this should be /Z7 /O0 minimal-yet-existent debugging help. All targets should always produce debugging data, generally detached from the envoy-static binary output.

However, it causes us to lose our support for .pdb files. But bazel seems sadly ignorant of both gcc linker debug-info splitting and windows pdb output. It seems like we need some cohesive solution to both problems directly from the rules_foreign_cc subproject and primary bazel project.

I've vetted all three build targets and all cooperate as well as they did with specific targets, after a bit of munging of resulting library names and commenting out the .pdb targets. I think we should proceed with this angle of attack on the challenge, and address the windows .pdb and linux objcopy splitting as another matter.

Reply:
@hlopko , @irengrig

We would propose adding this commit to this patch;
https://github.com/greenhouse-org/envoy/commit/d421a2d650c2a6ac24924ef4a019ed03ef7c36b4

This patch drops all the static toggles and drops the use of _RELEASE/_DEBUG/_RELWITHDEBINFO logic in CMake to exclusively use flags passed in by Bazel.

It has one significant drawback; it becomes necessary for us to add the .pdb logic to the foreign_cc_rules. Here are my observations on implementing this for bazel_foreign_cmake targets;

Effectively, where ever we have a static library /OUT: target, at compile-time we would add the compiler flag /Fd:{libname}.pdb for compiling each .obj file. Where ever we have a binary (exe or dll) /OUT: target, we would add the linker flag /PDB:{libname}.pdb.

From a Bazel-perspective, it is always helpful to have optional (discardable) .pdb output, much like linux debuginfo packages. For fastbuild, we want only the most trivial (quickest) debug info (/Z7), for real releases or debugging we want to be able to drill in deeper (/Zi). For fastbuild we don't want to waste cycles on link-time optimizations, for releases we do.

static lib object: compile /Fd:{libname}.pdb
dynamic lib/binary object: link /PDB:{libname}.pdb  [discarding temporary /Fd output]

fastbuild;  compile /O0 /Z7, link /DEBUG /INCREMENTAL:YES /OPT:NOREF
dbg:  compile /O0 /Zi, link /DEBUG /INCREMENTAL:NO /OPT:NOREF
opt:  compile /O2 /Zi, link /DEBUG /INCREMENTAL:NO /OPT:REF

I'm sure there are other small tweaks to consider to observe the spirit of the Bazel fastbuild/dbg/opt pattern.

We would appreciate your feedback on this approach, and whether (short-term) we can live without .pdb file support in favor of cleaning up the cmake logic now.

Reply:
@lizan @irengrig @hlopko are there any remaining questions on this approach, and is this solution acceptable to you all (as presented in this pull request at present)?